src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java

```diff
@@ -281,8 +281,17 @@ public void onSubscribe(Flow.Subscription subscription) {
 
         @Override
         public void onNext(List<ByteBuffer> items) {
+            long expectedSize = Utils.remaining(items);
             try {
-                out.write(items.toArray(Utils.EMPTY_BB_ARRAY));
+                long bytesWritten = 0;
+                ByteBuffer[] buffers = items.toArray(Utils.EMPTY_BB_ARRAY);
+                while (bytesWritten < expectedSize) {
+                    long n = out.write(buffers);
+                    if (n <= 0) {
+                        throw new IOException("zero bytes written");
```

Comment:
"Returns: The number of bytes written, possibly zero"

Sorry for being dumb, but why is the returned zero bytes treated as an error? If I were to write a fuzzifier for that class, it'd be my first thing on the to do list to randomly return zero bytes - if the contract allows it, it should be handled gracefully.

Reply:
Well this has the potential of creating a busy loop since bytesWritten will not increase if zero is returned.

Reply:
I know that. But write should eventually make progress - even if it spins a few times. What's the point of having that JDK method explicitly state "may return 0" and then disregarding that as something that is an exceptional scenario? I don't see the consistency here... If it can return zero, fine - handle it by retrying. If this code falls into a spin loop and never makes progress - this would be a problem somewhere else (perhaps in the implementation of the file channel), not here.

Reply:
I'm on the fence a little on this point. I really dislike the potential for a busy loop, but also don't want the added complication of retry/backoff in this loop. But really the latter is the right solution. lemme see how it looks.

Reply:
Don't get me wrong - you folks are much deeper into low-level details than I am so perhaps it isn't possible that the OS or the hardware ever returns 0 here. From the public API contract perspective there's nothing preventing it from happening though. It's a bit strange for me to see the internal implementation accepts a stricter requirement here.

Reply:
I added a trivial retry mechanism see https://github.com/openjdk/jdk/pull/11722/commits/c1355274911537fd25eae873908b84617bbb2e58